presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosAuthentication.java

```diff
@@ -0,0 +1,107 @@
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosPrincipal;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.login.Configuration;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.nio.file.Files.exists;
+import static java.nio.file.Files.isReadable;
+import static java.util.Collections.emptySet;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.security.SecurityUtil.getServerPrincipal;
+
+public class KerberosAuthentication
+{
+    private static final String KERBEROS_LOGIN_MODULE = "com.sun.security.auth.module.Krb5LoginModule";
+
+    private final KerberosPrincipal principal;
+    private final Configuration configuration;
+
+    public KerberosAuthentication(String principal, String keytabLocation)
+    {
+        requireNonNull(principal, "principal is null");
+        requireNonNull(keytabLocation, "keytabLocation is null");
+        Path keytabPath = Paths.get(keytabLocation);
+        checkArgument(exists(keytabPath), "keytab does not exists: " + keytabLocation);
+        checkArgument(isReadable(keytabPath), "keytab is not readable: " + keytabLocation);
+        this.principal = createKerberosPrincipal(principal);
+        this.configuration = createConfiguration(principal, keytabLocation);
+    }
+
+    public Subject getSubject()
+    {
+        Subject subject = new Subject(false, ImmutableSet.of(principal), emptySet(), emptySet());
+        try {
+            LoginContext loginContext = new LoginContext("", subject, null, configuration);
+            loginContext.login();
+            return loginContext.getSubject();
+        }
+        catch (LoginException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private static KerberosPrincipal createKerberosPrincipal(String principal)
+    {
+        try {
+            return new KerberosPrincipal(getServerPrincipal(principal, InetAddress.getLocalHost().getHostName()));
+        }
+        catch (IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private static Configuration createConfiguration(String principal, String keytabLocation)
+    {
+        Map<String, String> options = ImmutableMap.<String, String>builder()
+                .put("useKeyTab", "true")
+                .put("storeKey", "true")
+                .put("doNotPrompt", "true")
+                .put("isInitiator", "true")
+                .put("principal", principal)
+                .put("keyTab", keytabLocation)
```

Comment:
As we discussed we have product tests on our internal branch that do the `SELECT * FROM kerberized_hive.nation` from CLI authenticated with Kerberos with different principals. We looking forward to make those tests public and run them on Travis, but to do that we need #4877 merged as well as new version of `Tempto` and `Airlift` released.

However as we discussed Airlift contains `http.authentication.krb5.config` property, which is equivalent of ` java.security.krb5.conf` java property which is JVM wide. We have to consider what to do? Should we `http.authentication.krb5.config` property to some more generic name?